sherpa/astro/ui/serialize.py

```diff
@@ -984,65 +1111,179 @@ def _save_models(state, fh=None):
+    #
+    _remove_banner(out, orig_pos)
+
 
-def _save_xspec(fh=None):
-    """Save the XSPEC settings, if the module is loaded.
+def _save_xspec(out: OutType) -> None:
+    """Save the XSPEC settings.
 
     Parameters
     ----------
-    fh : None or file-like
-       If ``None``, the information is printed to standard output,
-       otherwise the information is added to the file handle.
+    out : dict
+       The output state
     """
 
-    if not hasattr(sherpa.astro, "xspec"):
+    # This case shpuld not happen, but just in case.
+    #
+    if xspec is None:
         return
 
-    # TODO: should this make sure that the XSPEC module is in use?
-    #       i.e. only write these out if an XSPEC model is being used?
-    _output("\n######### XSPEC Module Settings\n", fh)
-    xspec_state = sherpa.astro.xspec.get_xsstate()
-
-    cmd = "set_xschatter(%d)" % xspec_state["chatter"]
-    _output(cmd, fh)
-    cmd = 'set_xsabund("%s")' % xspec_state["abund"]
-    _output(cmd, fh)
-    cmd = "set_xscosmo(%g, %g, %g)" % (xspec_state["cosmo"][0],
-                                       xspec_state["cosmo"][1],
-                                       xspec_state["cosmo"][2])
-    _output(cmd, fh)
-    cmd = 'set_xsxsect("%s")' % xspec_state["xsect"]
-    _output(cmd, fh)
+    _output_banner(out, "XSPEC Module Settings")
+    xspec_state = xspec.get_xsstate()
+
+    chatter =  xspec_state["chatter"]
+    abund = xspec_state["abund"]
+    xsect = xspec_state["xsect"]
+    cs = xspec_state["cosmo"]
+    _output(out, f"set_xschatter({chatter})")
+    _output(out, f'set_xsabund("{abund}")')
+    _output(out, f"set_xscosmo({cs[0]:g}, {cs[1]:g}, {cs[2]:g})")
+    _output(out, f'set_xsxsect("{xsect}")')
 
     def tostatement(key, val):
-        return 'set_xsxset("{}", "{}")'.format(key, val)
+        return f'set_xsxset("{key}", "{val}")'
+
+    _save_entries(out, xspec_state["modelstrings"], tostatement)
+
+
+def _save_dataset_file(out: OutType, idstr: str, dset: Data) -> None:
+    """The data can be read in from a file."""
+
+    # TODO: this does not handle options like selecting the columns
+    #       from a file, or the number of columns.
+    #
+    ncols = None
+    if isinstance(dset, DataPHA):
+        dtype = 'pha'
+    elif isinstance(dset, DataIMG):
+        dtype = 'image'
+    else:
+        dtype = 'data'
+
+        # Can we estimate what ncols should be? This is a heuristic as
+        # it does not cover all cases.
+        #
+        if dset.staterror is not None:
+            # assume we read in independent axis/es, dependent axis,
+            # and then staterror.
+            ncols = len(dset.get_indep()) + 2
+
+    cmd = f'load_{dtype}({idstr}, "{dset.name}"'
+    if ncols is not None:
+        cmd += f", ncols={ncols}"
+
+    cmd += ")"
+    _output(out, cmd)
 
-    _save_entries(xspec_state["modelstrings"], tostatement, fh)
 
+def _output_wcs_import(out: OutType) -> None:
+    """Import the WCS symbol if not done already.
 
```

Comment:
```suggestion
    """Import the WCS symbol.

```
There is no test if it's done already. But there is also little harm in putting out that import statement more thna once.

Reply:
But there is a test (implicitly) because `out["imports"]` is a set, so it only gets added once. So I'm keeping the original text.